File: fft/go/fft/complex.go
package fft

type Complex struct {
	Real, Imag float64
}

func (c Complex) Add(other Complex) Complex {
	return Complex{c.Real + other.Real, c.Imag + other.Imag}
}

func (c Complex) Sub(other Complex) Complex {
	return Complex{c.Real - other.Real, c.Imag - other.Imag}
}

func (c Complex) Mul(other Complex) Complex {
	return Complex{c.Real*other.Real - c.Imag*other.Imag, c.Real*other.Imag + c.Imag*other.Real}
}

func (c Complex) MulScalar(scalar float64) Complex {
	return Complex{c.Real * scalar, c.Imag * scalar}
}

File: fft/swift/Sources/FFT/Complex.swift
public struct Complex {
  public var real: Double
  public var imag: Double

  public init(real: Double, imag: Double) {
    self.real = real
    self.imag = imag
  }
}

extension Complex {
  public static func + (l: Complex, r: Complex) -> Complex {
    return Complex(real: l.real + r.real, imag: l.imag + r.imag)
  }

  public static func - (l: Complex, r: Complex) -> Complex {
    return Complex(real: l.real - r.real, imag: l.imag - r.imag)
  }

  public static func * (l: Complex, r: Complex) -> Complex {
    return Complex(
      real: l.real * r.real - l.imag * r.imag,
      imag: l.real * r.imag + l.imag * r.real)
  }

  public static func * (l: Complex, r: Double) -> Complex {
    return Complex(real: l.real * r, imag: l.imag * r)
  }
}
